File: src/main/java/io/github/starwishsama/namelessbot/config/BackupHelper.java
package io.github.starwishsama.namelessbot.config;

import cn.hutool.core.io.file.FileWriter;
import com.google.gson.Gson;
import io.github.starwishsama.namelessbot.BotConstants;
import io.github.starwishsama.namelessbot.BotMain;

import java.io.File;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class BackupHelper {
    private static final File backupLocation = new File(BotMain.getJarPath() + "/backups");

    public static void createBackup(){
        if (!backupLocation.exists()){
            backupLocation.mkdirs();
        }

        String fileName = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ss"));

        FileWriter.create(new File(backupLocation, fileName + ".json")).write(new Gson().toJson(BotConstants.users));
        BotMain.getLogger().log("[备份] 自动备份完成");
    }
}

File: src/main/kotlin/io/github/starwishsama/comet/file/BackupHelper.kt
package io.github.starwishsama.comet.file

import cn.hutool.core.io.file.FileWriter
import io.github.starwishsama.comet.BotVariables
import io.github.starwishsama.comet.utils.FileUtil
import io.github.starwishsama.comet.utils.TaskUtil
import java.io.File
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.concurrent.TimeUnit

object BackupHelper {
    private val location: File = FileUtil.getChildFolder("backups")
    private val dateFormatter: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ss")

    private fun createBackup(){
        try {
            if (!location.exists()) {
                location.mkdirs()
            }

            val backupTime = LocalDateTime.now()
            val backupFileName = "backup-${dateFormatter.format(backupTime)}.json"

            val backupFile = File(location, backupFileName)
            backupFile.createNewFile()
            FileWriter.create(backupFile, Charsets.UTF_8)
                    .write(BotVariables.gson.toJson(BotVariables.users))
            BotVariables.logger.info("[备份] 备份成功! 文件名是 $backupFileName")
        } catch (e: Exception) {
            BotVariables.logger.error("[备份] 尝试备份时发生了异常", e)
        }
    }

    fun scheduleBackup() =
        TaskUtil.runScheduleTaskAsync(0, BotVariables.cfg.autoSaveTime, TimeUnit.MINUTES, BackupHelper::createBackup)
}